snuba/consumers/consumer.py

```diff
@@ -441,26 +349,52 @@ class MultistorageKafkaPayload(NamedTuple):
     payload: KafkaPayload
 
 
+def process_message(
+    processor: MessageProcessor, message: Message[KafkaPayload]
+) -> Union[None, JSONRowInsertBatch, ReplacementBatch]:
+    result = processor.process_message(
+        rapidjson.loads(message.payload.value),
+        KafkaMessageMetadata(
+            message.offset, message.partition.index, message.timestamp
+        ),
+    )
+
+    if isinstance(result, InsertBatch):
+        return JSONRowInsertBatch(
+            [json_row_encoder.encode(row) for row in result.rows],
+            result.origin_timestamp,
+        )
+    else:
+        return result
+
+
 def process_message_multistorage(
-    message: Message[MultistorageKafkaPayload],
+    storage_keys: Sequence[StorageKey], message: Message[KafkaPayload],
 ) -> Sequence[Tuple[StorageKey, Union[None, JSONRowInsertBatch, ReplacementBatch]]]:
     # XXX: Avoid circular import on KafkaMessageMetadata, remove when that type
     # is itself removed.
     from snuba.datasets.storages.factory import get_writable_storage
 
-    value = rapidjson.loads(message.payload.payload.value)
+    storages: MutableSequence[WritableTableStorage] = []
+
+    for storage_key in storage_keys:
+        storage = get_writable_storage(storage_key)
+        filter = storage.get_table_writer().get_stream_loader().get_pre_filter()
+        if filter is None or not filter.should_drop(message):
+            storages.append(storage)
+
```

Comment:
Why this change ?

Reply:
I wanted to reuse the strategy factory code here. This meant converting the multistorage strategy from `transform -> filter -> transform -> collect` into `transform -> collect`. So now this transform step includes the parts that were here before https://github.com/getsentry/snuba/blob/master/snuba/consumers/consumer.py#L617-L620

Reply:
This opens up an issue.
Before we were doing this filtering step first
```
return TransformStep(
    self.__find_destination_storages,
    FilterStep(self.__has_destination_storages, strategy,),
)
```
that was the very first thing that would happen before going to the parallel step.
For a consumer like cdc, that is multi storage, but that drops the vast majority of the messages, that meant that before we would be dropping most messages before getting into the dispatch step of the parallel transformation (the dispatch is a heavy one in the parallel consumer as it can be slowed down by lock contention when the parallel step is trivial. Most of the time would be spent in filling batches and waiting for them to be ready).
With your change this filtering is in `process_message_multistorage`, which is the parallel step and thus happens  AFTER dispatching. 

I would be concerned this would have a negative impact on the throughput of consumers like cdc.

Can we move this back to the filter ?
Also maybe leaving a note about this issue in a comment would help the next time somebody approaches this code. It can be tricky to notice at first.

Reply:
Damn I did not realise CDC was using this strategy. Let me think about how to keep this as a filter step.

Reply:
Since you are there. Could you please ensure you test these cases manually as we have little coverage of the consumer end to end:
1) Run a multiprocess consumer on one storage alone and ensure it consume in parallel
2) Same without switching the consumer to multi processing
3) Run a multi storage consumer with 1 process (Starting devserves now starts the cdc consumer in multi storage mode. Then just enable cdc in sentry)
4) Run a multi storage consumer with 2-3 processes

Reply:
I've tried the 4 combinations (with the events and errors storage for multistorage consumer) and it works. Still trying to get CDC running in Sentry though.

Reply:
Also added the multistorage filter step back here. Note we still need to re-filter in process_message_multistorage though to ensure that the event is only written to the right storage. If you think that this double filtering is too undesirable I can revert these changes and not inherit from the generic strategy factory.

Reply:
Running the filter twice is not dangerous. Though it will be quite confusing for those who look at the code.
It seems to me this issue is actually a signal that we are trying to abstract together a few things that should not be abstracted together.
here we have serial and parallel transform on one dimension. Plus single storage and multi storage on the other dimension.
The abstraction works well for serial transformations. You can easily represent the process as filter-proess-collect. For parallel consumption this is transform-parallel_transform-collect instead.
What if we captured that difference (the two types of process) in the type parameters? Which basically requires two options:
- having a StandardConsumerStrategyFactory and a parallel one. In one the first step is a fitler in the other it is a transformation.
OR
- Turn the  ConsumerStrategyFactory into a process with three steps: pre-transform (instead of filter) transform and collect. I like the other option better since, in this case, the need for two transform step is due to an implementation detail we are trying to hide (and which cannot really be hidden).

Reply:
I reverted the changes to the multistorage consumer; no more fitting this into the generic 3 steps, we're back to 4. I don't think it's worth providing something in the library to handle this particular 4 step scenario as I don't imagine
it's a common one.

Personally I thought the earlier approach was more straightforward because there wasn't the extra intermediate format that the message was transformed into to think about, but this may be highly subjective.